shared/lib_battery_dispatch_pvsmoothing_fom.cpp

```diff
@@ -0,0 +1,440 @@
+
+    ssc_number_t battery_energy = _Battery->energy_nominal(); // check units in equations below 
+    ssc_number_t battery_power = m_batteryPower->powerBatteryChargeMaxAC;
+    // scale by nameplate per ERPI code
+    battery_energy = m_batt_dispatch_pvs_nameplate_ac > 0 ? battery_energy / m_batt_dispatch_pvs_nameplate_ac : battery_energy;
+    battery_power = m_batt_dispatch_pvs_nameplate_ac > 0 ? battery_power / m_batt_dispatch_pvs_nameplate_ac : battery_power;
+
+
+    //#local variables
+    ssc_number_t previous_power = 0;
+    ssc_number_t battery_soc = m_batt_dispatch_pvs_initial_SOC/ 100.0; 
+    m_batt_dispatch_pvs_outpower_vec.clear();
+    m_batt_dispatch_pvs_battpower_vec.clear();
+    m_batt_dispatch_pvs_battsoc_vec.clear();
+    m_batt_dispatch_pvs_curtail_vec.clear();
+    m_batt_dispatch_pvs_violation_list_vec.clear();
+
+    m_batt_dispatch_pvs_outpower_vec.reserve(nRecordsSampled);
+    m_batt_dispatch_pvs_battpower_vec.reserve(nRecordsSampled);
+    m_batt_dispatch_pvs_battsoc_vec.reserve(nRecordsSampled);
+    m_batt_dispatch_pvs_curtail_vec.reserve(nRecordsSampled);
+    m_batt_dispatch_pvs_violation_list_vec.reserve(nRecordsSampled);
+
+
+    // output accumulators
+    size_t violation_count = 0;
+    ssc_number_t total_energy = 0;
+
+    if (!en_forecast)
+        kf = 0;
+
+    //  #iterate through time - series
+    for (size_t ndx_sampled = 0; ndx_sampled < nRecordsSampled; ndx_sampled++) {
+        ssc_number_t pv_power = m_pv_power_input_sampled_vec[ndx_sampled];
+        ssc_number_t out_power = 0;
+        ssc_number_t battery_power_terminal = 0;
+        ssc_number_t forecast_power = 0;
+        forecast_power = m_forecast_pv_energy_vec[ndx_sampled];
+        //                #calculate controller error
+        ssc_number_t delta_power = pv_power - previous_power; //#proportional error
+        ssc_number_t soc_increment = battery_soc + (pv_power - previous_power) * power_to_energy_conversion_factor;// #integral error
+        ssc_number_t future_error = previous_power * forecast_shift_periods * power_to_energy_conversion_factor - forecast_power; //#derivitive error
+        ssc_number_t error = kp * delta_power + ki * (soc_increment - soc_rest * battery_energy) - kf * future_error;
+
+        //                #calculate the desired output power, enforce ramp rate limit
+        if (error > 0)
+            out_power = previous_power + std::min(max_ramp, std::abs(error));
+        else
+            out_power = previous_power - std::min(max_ramp, std::abs(error));
+
+        //            #enforce grid power limits
+        if (AC_upper_bound_on) {
+            if (out_power > AC_upper_bound)
+                out_power = AC_upper_bound;
+        }
+        if (AC_lower_bound_on) {
+            if (out_power < AC_lower_bound)
+                out_power = AC_lower_bound;
+        }
+
+        //                  #calculate desired(unconstrained) battery power
```

Comment:
Can these be reformatted as C++ comments? The whitespace is hard to read.

Reply:
Yes, they were originally taken directly from the Python code from EPRI

Reply:
Resolved with last commit